quantum/repeat_key.c

```diff
@@ -0,0 +1,282 @@
+    last_record         = *record;
+    last_record.keycode = keycode;
+    last_repeat_count   = 0;
+}
+
+void repeat_key_invoke(const keyevent_t* event) {
+    // It is possible (e.g. in rolled presses) that the last key changes while
+    // the Repeat Key is pressed. To prevent stuck keys, it is important to
+    // remember separately what key record was processed on press so that the
+    // the corresponding record is generated on release.
+    static keyrecord_t registered_record       = {0};
+    static int8_t      registered_repeat_count = 0;
+    // Since this function calls process_record(), it may recursively call
+    // itself. We return early if `processing_repeat_count` is nonzero to
+    // prevent infinite recursion.
+    if (processing_repeat_count || !last_record.keycode) {
+        return;
+    }
+
+    if (event->pressed) {
+        update_last_repeat_count(1);
+        // On press, apply the last mods state, stacking on top of current mods.
+        register_weak_mods(last_mods);
+        registered_record       = last_record;
+        registered_repeat_count = last_repeat_count;
+    }
+
+    // Generate a keyrecord and plumb it into the event pipeline.
+    registered_record.event = *event;
+    processing_repeat_count = registered_repeat_count;
+    process_record(&registered_record);
+    processing_repeat_count = 0;
+
+    // On release, restore the mods state.
+    if (!event->pressed) {
+        unregister_weak_mods(last_mods);
+    }
+}
+
+#ifndef NO_ALT_REPEAT_KEY
+/**
+ * @brief Find alternate keycode from a table of opposing keycode pairs.
+ * @param table Array of pairs of basic keycodes, declared as PROGMEM.
+ * @param table_size_bytes The size of the table in bytes.
+ * @param target The basic keycode to find.
+ * @return The alternate basic keycode, or KC_NO if none was found.
+ *
+ * @note The table keycodes and target must be basic keycodes.
+ *
+ * This helper is used several times below to define alternate keys. Given a
+ * table of pairs of basic keycodes, the function finds the pair containing
+ * `target` and returns the other keycode in the pair.
+ */
+static uint8_t find_alt_keycode(const uint8_t (*table)[2], uint8_t table_size_bytes, uint8_t target) {
+    const uint8_t* keycodes = (const uint8_t*)table;
+    for (uint8_t i = 0; i < table_size_bytes; ++i) {
+        if (target == pgm_read_byte(keycodes + i)) {
+            // Xor (i ^ 1) the index to get the other element in the pair.
+            return pgm_read_byte(keycodes + (i ^ 1));
+        }
+    }
+    return KC_NO;
+}
+
+uint16_t get_alt_repeat_key_keycode(void) {
+    uint16_t keycode = last_record.keycode;
+    uint8_t  mods    = last_mods;
+
+    // Call the user callback first to give it a chance to override the default
+    // alternate key definitions that follow.
+    uint16_t alt_keycode = get_alt_repeat_key_keycode_user(keycode, mods);
+
+    if (alt_keycode) {
+        return alt_keycode;
+    }
```

Comment:
This does not seem to offer a way to make a predefined alternate key do nothing (apart from returning a keycode which is not handled anywhere, but that would still result in calls to `process_record()`).

Maybe it would be better to redefine the return value of `get_alt_repeat_key_keycode_user()` as:
- `KC_NO` — do nothing (any predefined alternate key is not used);
- `KC_TRNS` — use the predefined alternate key if it exists;
- anything else — use the specified keycode.

```suggestion
    // Call the user callback first to give it a chance to override the default
    // alternate key definitions that follow.
    uint16_t alt_keycode = get_alt_repeat_key_keycode_user(keycode, mods);

    if (alt_keycode != KC_TRANSPARENT) {
        return alt_keycode;
    }
```